### ISSUE TITLE
Implementing more realistic transforms in Robust DPatch EoT

### ISSUE THREAD

**Issue by mut0-mitre** [Trusted-AI/adversarial-robustness-toolbox#1907]
**Is your feature request related to a problem? Please describe.**
Robust DPatch has many use cases, as my team has used numerous times for our research. However, the current implementation provides limited type and range of transformations in its EoT. Also, the current implementation transforms whole patched image as opposed to transforming patch before applying onto the image. For real world use cases that my team has seen, we would like 1. more diverse transforms and 2. transforming the patch before applying onto image.

**Describe the solution you'd like**
I have implemented solution for the problem above, by defining another class called `EoTRobustDPatch` and using transform methods defined in [adversarial_patch_pytorch](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/dev_1.13.0/art/attacks/evasion/adversarial_patch/adversarial_patch_pytorch.py). It has been verified and used in my team's experiments.

**Describe alternatives you've considered**
My team considered simply tweaking the source code, but we found that it required substantial enough changes that we would hinder/conflict with ART development, so we decided to define a new attack that can be simply added to the current library of attacks.

**Additional context**
The code is currently under release process in my team. Once the code is approved, I will be able to push the code and make my pull request ready. If there are any requests, questions, or comments regarding this issue, please let me know. Thank you!

**Comment by beat-buesser** [comment]
Hi @mut0-mitre Thank you very much for proposing a new feature for ART! Have you considered as alternative the existing combination of `AdversarialPatchPyTorch` plus an estimator with EoT preprocessors defined in `art.preprocessing.expectation_over_transformation`?

**Comment by mut0-mitre** [comment]
Hi @beat-buesser, I have not considered using `AdversarialPatchPyTorch`+`art.preprocessing.expectation_over_transformation`, as my focus has been `RobustDPatch`. Also, I was looking for more diverse set of transforms than those defined in `art.preprocessing.expectation_over_transformation`. `AdversarialPatchPyTorch` seems to offer wide range and variety of transforms, leading me to combine `RobustDPatch`+`AdversarialPatchPyTorch.random_overlay()`. I am open for any feedback though. What do you think?

**Comment by beat-buesser** [comment]
Hi @mut0-mitre I think we had a similar idea for ART 1.10.0 where we extended `AdversarialPatchPyTorch` to support object detection and addition of PGD L-Inf optimization to provide a PyTorch-specific replacement for `RobustDPatch` which can provide flexible configuration of EoTs with `art.preprocessing.expectation_over_transformation`. This code outlines how to to run `RobustDPatch` with `AdversarialPatchPyTorch`:

```python
from art.attacks.evasion import AdversarialPatchPyTorch
from art.estimators.object_detection import PyTorchFasterRCNN
from art.preprocessing.expectation_over_transformation import (
        EoTImageCenterCropPyTorch,
        EoTImageRotationPyTorch,
        EoTBrightnessPyTorch,
    )

eots = []

eots.append(
    EoTImageCenterCropPyTorch(
        nb_samples=2,
        clip_values=(0, 255),
        size=4,
        label_type="object_detection",
        apply_fit=False,
        apply_predict=True,
    )
)
eots.append(
    EoTImageRotationPyTorch(
        nb_samples=2,
        clip_values=(0, 255),
        angles=[0, 180],
        label_type="object_detection",
        apply_fit=False,
        apply_predict=True,
    )
)
eots.append(
    EoTBrightnessPyTorch(
        nb_samples=2,
        clip_values=(0, 255),
        delta=3,
        apply_fit=False,
        apply_predict=True,
    )
)

frcnn = PyTorchFasterRCNN(
    clip_values=(0, 255),
    attack_losses=("loss_classifier", "loss_box_reg", "loss_objectness", "loss_rpn_box_reg"),
    channels_first=False,
    preprocessing_defences=eots,
    device_type="gpu",
)

attack = AdversarialPatchPyTorch(
    estimator=frcnn,
    rotation_max=0.0,
    scale_min=1,
    scale_max=1,
    distortion_scale_max=0.0,
    learning_rate=1.0,
    max_iter=2,
    batch_size=16,
    patch_shape=(3, 100, 100),
    patch_location=(50, 50),
    patch_type="square",
    optimizer="pgd",
    targeted=True,
    summary_writer=True,
    verbose=True,
)

patch, _ = attack.generate(x=image, y=predictions)
````

There are additional lines of code required to set-up the EoTs, but this also enables the user to add/implement new EoTs, remove or reorder them. Plus the EoTs can be reused in other attacks.

I think it could be interesting to compare the performance of your code with the approach above?

**Comment by mut0-mitre** [comment]
@beat-buesser Thanks for the suggestion. Few questions about the snippet you provided above:
- Is the transform done on the whole background image, patched image, or the patch itself only?
- Can `AdversarialPatchPyTorch` provide additional transforms, or do all the parameters have to be set to 0 when using it with EoT `preprocessing`? (so for example, instead of rotation_max = 0, can we set rotation_max=nonzero)

Currently, I see few major differences between your code snippet and our code implementation. However, I believe that whether our implementation will benefit the ART community depends on people's use cases and needs. I might be making some assumptions so correct me if I am wrong, but the differences I see are:
- Our implementation allows continuous range of values for `rotation`, instead of multiples of 90 degrees
- Our implementation does transforms on the patch itself, instead of the whole image - this was done simply to meet our use case
- Our implementation allows perspective transform, or for that matter, any transforms defined in `AdversarialPatchPyTorch`.

Since the two implementations are different, it would be difficult to compare the performance. We could, however, characterize and document the differences, if desired. I am happy to move in any directions so long as it helps the community, though my team is looking for ways to contribute our implementation. Let me know your thoughts.

**Comment by beat-buesser** [comment]
Hi @mut0-mitre 

> Is the transform done on the whole background image, patched image, or the patch itself only?

In ART the EoT tools are applied to the image/input just before they are provided to the model (e.g for prediction, gradient calculation, etc.). In the case of `AdversarialPatchPyTorch` this means that the patch is first applied by the attack and before the patched image is provided to the model all EoT steps are applied by the ART estimator containing the model.

> Can AdversarialPatchPyTorch provide additional transforms, or do all the parameters have to be set to 0 when using it with EoT preprocessing? (so for example, instead of rotation_max = 0, can we set rotation_max=nonzero)

Yes, `AdversarialPatchPyTorch` provides all the transformations listed in its arguments. These transformations are used to sample different configurations of the patch before it is applied to the image. The values are set to 0 because I have copied the code from a debugging script, these values can be non-zero.

> Our implementation allows continuous range of values for rotation, instead of multiples of 90 degrees

Are you referring to the rotation applied to the patch or the patched image? `AdversarialPatchPyTorch` provides continuous range of values for rotation angles for the patch. `EoTImageRotationPyTorch` provides rotation of multiples of 90 degrees for the patched images.

> Our implementation does transforms on the patch itself, instead of the whole image - this was done simply to meet our use case

`AdversarialPatchPyTorch` applies transforms the patch before applying it to the image using `AdversarialPatchPyTorch._random_overlay()`

> Our implementation allows perspective transform, or for that matter, any transforms defined in AdversarialPatchPyTorch

`AdversarialPatchPyTorch` also applies perspective transform as well as all transforms defined in `AdversarialPatchPyTorch` because it is the same attack, and also from where you got `AdversarialPatchPyTorch._random_overlay()`.

I think the existing tools and your implementation are trying to do the same. The reason behind the existing ART tools mentioned here is that `AdversarialPatchPyTorch` contains all patch specific sampling because it is the main patch attack in ART, while the tools in `art.preprocessing.expectation_over_transformation` provide EoT sampling which is applied to the eventual input to a model. These EoT are in a separate module to allow reusing them across all applications of the respective ART estimator. With this we also would like to allow the user of ART to apply new EoTs without having to change the code of an attack by adding additional EoTs to the estimator as shown in the code snippet above.

I think it would be amazing if you and your team could become contributors to ART and you would be very welcome. At the moment my biggest concern is that you might have started working on a feature which probably already exist before proposing and discussing it here. If you like I can take a closer look at the code in your repo?

**Comment by mut0-mitre** [comment]
Hi @beat-buesser, thanks for addressing all the questions in detail! Looks like ART does have a pretty comprehensive set of transforms for `AdversarialPatchPyTorch`. Just one more important question - does `AdversarialPatchPyTorch` support object detection estimator? From a quick look at the documentation and code, it looks like it's designed for classifier. 

Right, I probably should have started this discussion before beginning our development...In the future, if we come up with development needs again, we will definitely consult early. 

I have not been able to post the code yet, since we have internal review/release process that needs to be completed. We are unfortunately still early in this release process...

**Comment by beat-buesser** [comment]
Hi @mut0-mitre

> does `AdversarialPatchPyTorch` support object detection estimator?

Yes, it does, but it is currently missing the correct type description for the estimator, it should include object detectors and we need to fix it soon.

I think there are many items where we could use your help and maybe include some of your codes. One of our goals with EoT and `AdversarialPatchPyTorch` was aiming at speed improvements by running everything (EoT, attack, model) on GPU. This calculates all EoT samples in parallel, which can be limited by available memory on GPU. An advantage of the `DPatch` approach is that it loops over the EoT samples in sequence with a for-loop. It could be interesting to see if `AdversarialPatchPyTorch` could be extended to run EoT in parallel if memory allows it, and automatically fall back to sequential sampling otherwise.

It also sounds like you and your team have done a lot of experiments with patches against object detectors. I think it would be very interesting to compare the attack success of both approaches to find bugs, improvements, etc.
